lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3771,79 +3642,1099 @@ inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
+      }
+    }
+  }
+
+  SILValue emitZeroDirect(CanType type, SILLocation loc) {
+    auto builder = getDifferentialBuilder();
+    auto silType = getModule().Types.getLoweredLoadableType(
+        type, ResilienceExpansion::Minimal);
+    auto *buffer = builder.createAllocStack(loc, silType);
+    auto *initAccess = builder.createBeginAccess(
+        loc, buffer, SILAccessKind::Init, SILAccessEnforcement::Static,
+        /*noNestedConflict*/ true, /*fromBuiltin*/ false);
+    emitZeroIndirect(type, initAccess, loc);
+    builder.createEndAccess(loc, initAccess, /*aborted*/ false);
+    auto readAccess = builder.createBeginAccess(
+        loc, buffer, SILAccessKind::Read, SILAccessEnforcement::Static,
+        /*noNestedConflict*/ true, /*fromBuiltin*/ false);
+    auto *loaded = builder.createLoad(
+        loc, readAccess, getBufferLOQ(type, getDifferential()));
+    builder.createEndAccess(loc, readAccess, /*aborted*/ false);
+    builder.createDeallocStack(loc, buffer);
+    return loaded;
+  }
+
+  SILValue materializeTangentDirect(
+      AdjointValue val, SILLocation loc) {
+    assert(val.getType().isObject());
+    LLVM_DEBUG(getADDebugStream() <<
+               "Materializing tangents for " << val << '\n');
+    assert(val.getKind() != AdjointValueKind::Aggregate
+           && "Tangent values cannot be aggregated in forward mode.");
+    switch (val.getKind()) {
+      case AdjointValueKind::Zero: {
+        auto zeroVal = emitZeroDirect(val.getSwiftType(), loc);
+        return zeroVal;
+      }
+      // Aggregate case here is necessary until we update 'AdjointValue' to be
+      // 'TangentValue'. An assertion is made above to make sure we
+      // never see an Aggregate in forward mode until this change is made.
+      case AdjointValueKind::Aggregate:
+      case AdjointValueKind::Concrete:
+        return val.getConcreteValue();
+    }
+  }
+
+  SILValue materializeTangent(AdjointValue val, SILLocation loc) {
+    if (val.isConcrete()) {
+      LLVM_DEBUG(getADDebugStream()
+                 << "Materializing tangent: Value is concrete.\n");
+      return val.getConcreteValue();
+    }
+    LLVM_DEBUG(getADDebugStream() << "Materializing tangent: Value is "
+                                     "non-concrete. Materializing directly.\n");
+    return materializeTangentDirect(val, loc);
+  }
+
+  //--------------------------------------------------------------------------//
+  // Tangent buffer mapping
+  //--------------------------------------------------------------------------//
+
+  void setTangentBuffer(SILBasicBlock *origBB, SILValue originalBuffer,
+                        SILValue adjointBuffer) {
+    assert(originalBuffer->getType().isAddress());
+    auto insertion =
+        bufferMap.try_emplace({origBB, originalBuffer}, adjointBuffer);
+    assert(insertion.second); (void)insertion;
+  }
+
+  SILValue getTangentProjection(SILBasicBlock *origBB,
```

Comment:
Please remove unexercised copied code. This method is unnecessary for tangent emission.